### ISSUE TITLE
Issue in custom options

### ISSUE THREAD

**Issue by dev-ajaysaini** [magento/magento2#25581]
Getting issue on product page and cart page if product option type is changed from text to date after adding product to cart.
https://prnt.sc/pwckw1

### Preconditions (*)

Use Magento 2.3.3 instance

### Steps to reproduce (*)

1. Create simple product with a custom option with Option Type **Field**
2. Add this product to cart.
3. Again edit product and change Option Type to **Date**
4. Now check cart page and add the same product again to cart.

### Actual result (*)

1. Cart page is blank.
2. Getting error in console.
![image](https://user-images.githubusercontent.com/42237032/68760645-59a2d400-0638-11ea-8d76-cd78674aab3c.png)

![image](https://user-images.githubusercontent.com/42237032/68760607-44c64080-0638-11ea-9842-764eaf91fc60.png)

**Comment by krishprakash** [comment]
@magento give me 2.3-develop instance

**Comment by magento-engcom-team** [comment]
Hi @krishprakash. Thank you for your request. I'm working on Magento 2.3-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @krishprakash, here is your [Magento instance](https://i-25581-2-3-develop.instances.magento-community.engineering).
Admin access: [https://i-25581-2-3-develop.instances.magento-community.engineering/admin](https://i-25581-2-3-develop.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by magento-engcom-team** [comment]
@krishprakash Thank you for verifying the issue.

Unfortunately, not enough information was provided to acknowledge ticket. Please consider adding the following:
- [x] Add `"Component: "` label(s) to this ticket based on verification result. If uncertain, you may follow the best guess

Once all required information is added, please add label `"Issue: Confirmed"` again.
Thanks!

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @krishprakash 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-23089` were created

**Issue Available**: @krishprakash, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by jmcharadva** [comment]
@magento-engcom-team 

I have looked into this issue.

I determined that customer-data.js uses the localStorage within the browser itself to store a variety of information, so you can clear the Magento Cache all you want and you will still have issues.

TLDR:

The fix in Google Chrome is hit F12 to open the Developer Toolbar, then select the Application tab. Clear the Local Storage, Session Storage, and Cookies. Reload the page and all should be well. You will have to do this on your admin page as well.

**Comment by krzksz** [comment]
I followed the reproduction steps and this is not a frontend issue. When cart loads, invalidation of customer data is triggered properly and it is backend, which should handle the error case and return new cart section contents without products that have incorrect custom options.

**Comment by Usik2203** [comment]
I have investigated this issue and it is backed issue

When You add product which has custom option with type Field to cart, value of this option will be saved into  table `quote_item_option` 

After changing custom option type to Date and going to cart

Is expected to value in Date format. 

Actually  we have such flow

1. Add product 
http://i.imgur.com/ZbgSMfU.png

2. Option was saved in db 
http://i.imgur.com/UB4UsN1.png

3. Changed type from Field to Date and  expected string in Date format. 
 http://i.imgur.com/qDJcOn8.png

4. We are waiting date format string but get wrong format string for date
http://i.imgur.com/Vcniw01.png


Root point of exception is in this constructor https://www.php.net/manual/ru/datetime.construct.php

**Comment by dev-ajaysaini** [comment]
There is another issue at the time of reorder.
Buyer Not able to reorder products when custom option (custom option type) changed.
